File: sdks/python/agentcomms/resources/slack.py
# SPDX-License-Identifier: FSL-1.1-Apache-2.0
# © 2026 Victory. Licensed under the Functional Source License, Version 1.1,
# with Apache 2.0 Future License. See LICENSE for details.
"""Slack channel-native sub-surface resource."""
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from agentcomms.client import Client


class SlackWorkspaceChannels:
    def __init__(self, client: "Client", agent_id: str, team_id: str) -> None:
        self._client = client
        self._agent_id = agent_id
        self._team_id = team_id

    def _base(self) -> str:
        return f"/agents/{self._agent_id}/slack/workspaces/{self._team_id}/channels"

    def list(self) -> list[dict[str, Any]]:
        data = self._client._request("GET", self._base())
        return data.get("channels", [])

    def messages(self, channel_id: str, *, limit: int = 50) -> list[dict[str, Any]]:
        data = self._client._request("GET", f"{self._base()}/{channel_id}/messages", params={"limit": limit})
        return data.get("messages", [])

    def post(self, channel_id: str, *, body: str, blocks: Any = None) -> dict[str, Any]:
        payload: dict[str, Any] = {"body": body}
        if blocks:
            payload["blocks"] = blocks
        return self._client._request("POST", f"{self._base()}/{channel_id}/messages", json=payload)


class SlackWorkspace:
    def __init__(self, client: "Client", agent_id: str, team_id: str) -> None:
        self._client = client
        self._agent_id = agent_id
        self._team_id = team_id
        self.channels = SlackWorkspaceChannels(client, agent_id, team_id)

    def dm(self, user_id: str, *, body: str) -> dict[str, Any]:
        return self._client._request(
            "POST",
            f"/agents/{self._agent_id}/slack/workspaces/{self._team_id}/users/{user_id}/messages",
            json={"body": body},
        )


class SlackResource:
    def __init__(self, client: "Client", agent_id: str) -> None:
        self._client = client
        self._agent_id = agent_id

    def workspaces(self) -> list[dict[str, Any]]:
        data = self._client._request("GET", f"/agents/{self._agent_id}/slack/workspaces")
        return data.get("workspaces", [])

    def workspace(self, team_id: str) -> SlackWorkspace:
        return SlackWorkspace(self._client, self._agent_id, team_id)

File: sdks/node/src/resources/slack.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Victory (Intergalactic Tech).
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
import type { Client } from "../client.js";

export class SlackWorkspaceChannels {
  constructor(private client: Client, private agentId: string, private teamId: string) {}

  async list(): Promise<unknown[]> {
    const data = await this.client.request<{ channels: unknown[] }>(
      "GET",
      `/agents/${this.agentId}/slack/workspaces/${this.teamId}/channels`,
    );
    return data.channels ?? [];
  }

  async messages(channelId: string, params: { limit?: number } = {}): Promise<unknown[]> {
    const qs = params.limit ? `?limit=${params.limit}` : "";
    const data = await this.client.request<{ messages: unknown[] }>(
      "GET",
      `/agents/${this.agentId}/slack/workspaces/${this.teamId}/channels/${channelId}/messages${qs}`,
    );
    return data.messages ?? [];
  }

  async post(channelId: string, params: { body: string; blocks?: unknown }): Promise<unknown> {
    return this.client.request(
      "POST",
      `/agents/${this.agentId}/slack/workspaces/${this.teamId}/channels/${channelId}/messages`,
      { text: params.body, blocks: params.blocks },
    );
  }
}

export class SlackWorkspace {
  readonly channels: SlackWorkspaceChannels;

  constructor(private client: Client, private agentId: string, readonly teamId: string) {
    this.channels = new SlackWorkspaceChannels(client, agentId, teamId);
  }

  async dm(userId: string, params: { body: string }): Promise<unknown> {
    return this.client.request(
      "POST",
      `/agents/${this.agentId}/slack/workspaces/${this.teamId}/users/${userId}/messages`,
      { text: params.body },
    );
  }
}

export class SlackResource {
  constructor(private client: Client, private agentId: string) {}

  async workspaces(): Promise<unknown[]> {
    const data = await this.client.request<{ workspaces: unknown[] }>(
      "GET",
      `/agents/${this.agentId}/slack/workspaces`,
    );
    return data.workspaces ?? [];
  }

  workspace(teamId: string): SlackWorkspace {
    return new SlackWorkspace(this.client, this.agentId, teamId);
  }
}
